Test retries

palmerj3 opened the pull request:
## Summary

This PR introduces a new feature to Jest that allows you to re-run failed tests automatically n-times.
This is configurable via the Jest object.

Example:

`jest.retryTimes(3);
`
By default Jest will not re-run any failed tests.

The rationale behind this feature is that many teams use Jest for Integration and E2E tests. Higher level tests are much more likely to be flaky, and flakiness is often not the fault of the test itself (e.g. network problems).

This option provides the user a much better experience in CI since tests can be retried a few times before failing. The current behavior requires entire CI jobs to be re-triggered when a single test fails.

This PR also introduces a new field in TestResults for reporters to make use of. It is called "invocations" and represents the number of times a given test was executed.

**This functionality is only available if you use jest-circus!**

## Test plan

Added a e2e test to ensure that tests are only re-run when jest.retryTimes is set and jest-circus is used.

SimenB commented on packages/jest-cli/src/cli/args.js:
```diff
@@ -51,6 +51,17 @@ export const check = (argv: Argv) => {
 
+  const intRetries = parseInt(argv.testRetries, 10);
+  if (
+    (argv.hasOwnProperty('testRetries') && isNaN(intRetries)) ||
```
`Number.isNaN`

thymikee commented on packages/jest-circus/src/run.js:
```diff
@@ -28,30 +28,41 @@ import {
 const Promise = getOriginalPromise();
 
-const run = async (): Promise<RunResult> => {
+const run = async (numRetries: number): Promise<RunResult> => {
```
Can we stay consistent with the naming and use `testRetries` instead?

palmerj3 commented:
Great! I'll make these changes today and update the PR.

palmerj3 commented:
Also I have code working locally which updates the reporter API adding a new field for TestResults called invocations. I'm not sure how useful it would be so I haven't pushed it up. But I like the idea of a reporter being able to track which tests are retried since it's a signal about potential flakiness. Let me know your thoughts on that. I'm happy to include that if you think it's worthwhile.

SimenB commented:
Yeah seems useful. Can use it to inform about flakiness over time

aaronabramov commented:
i was also thinking about `test.retries(5, 'does stuff', () => {});` api to retry some nasty tests explicitly.
also it might make sense to reschedule these tests to the end of the file after other tests complete. e.g.:

```
✅test 1
❌test 2
✅test 3
✅test 4
---
✅test 2 (tetry)
```

the reason is that sometimes when you run e2e integration tests there's can be a spike in CPU usage (another thread decided to spawn something?) and the test times out. If you retry it right away the spike might be still going on and it'll fail again, but if you first try to execute everything else, see what passes, and then re-execute tests that failed you might have enough time for the CPU spike to die down.

aaronabramov commented:
also.. i noticed when executing fat integration tests, sometimes they timeout, jest moves on to the next test and the previous test eventually throws something. Jest treats it as a global error (because it does not know where it came from) and fails whatever current test is running, which results in a pretty confusing behavior.
With retries it will probably start throwing even more errors

palmerj3 commented:
@aaronabramov awesome! Is test.retry something you want me to implement as part of this or entirely to replace the CLI/Config-based approach?

Also regarding errors thrown after tests - not really sure the best way to make tests run later on. Push to the end of the test array? How would I do that?

aaronabramov commented:
@palmerj3 those we just my observations from last time i tried it! we can definitely iterate on those ideas later!
i don't mind CLI, but it kind of feels like it's too many config options (plus it won't be relevant for other runners)
maybe we should start by implementing it on the jest object? similar to `jest.setTimeout()`?
like
```js
jest.retryTimes(3);
```
or something like that

palmerj3 commented:
Awesome - you've given me a different perspective for approaching this problem. I'm going to try a few things out (jest obj, test.retry, etc) and see what comes of it. Thanks!

palmerj3 commented:
@aaronabramov I guess the thing I'm confused about is how I would responsibly go about retrying tests via the jest object? I can easily set a global to know how many retries are required in a given context but the act of actually retrying would have to be implemented by the test runner unless I override the test/it global and orchestrate from jest-runtime.

aaronabramov commented:
@palmerj3 yeah, i was thinking about exactly how `jest.setTimeout()` does it.
it sets the global value (that defaults to 1), that lives in jest-circus state and then once we're running tests, we access this value and check how many retries we have

palmerj3 commented:
Ok cool that's how I'll do it

palmerj3 commented:
@aaronabramov @SimenB if you get a chance please review again.

CI seems to be failing and I'm having a hard time debugging that... tests all pass locally on node 9 and node 10 so I'm not understanding why things fail on CI.

That said: I updated the PR so test retries are configured via the Jest object and I updated the reporter API so "invocations" field is available in TestResult.

Take a look when you can and offer ideas on how to fix CI :)

SimenB commented on packages/jest-runtime/src/index.js:
```diff
@@ -830,6 +830,11 @@ class Runtime {
     };
 
+    const retryTimes = (numTestRetries: number) => {
+      this._environment.global.retryTimes = numTestRetries;
```
can we put it on the `jest` object instead of as a global?

palmerj3 replied:
I'll see if that will work

palmerj3 replied:
@SimenB so I don't think that's going to work. From the context of jest-circus/run.js I don't have access to the jest object. Even if I did I think I would have to add both a getter and a setter for that to work. What do you think?

SimenB replied:
a clean (and private) way of speaking across that boundary would be awesome, but no need to tackle it at this moment 🙂

aaronabramov replied:
last time i did it with a `Symbol` that is attached to `global`. which is pretty much the same thing, but you only know it's there if you know what you're looking for :)

SimenB replied:
could we do that? Polluting the global namespace in user code seems icky

SimenB commented on e2e/__tests__/test_retries.test.js:
```diff
@@ -0,0 +1,60 @@
+    runJest(
+      'test-retries',
+      ['--config', JSON.stringify(reporterConfig), 'retry.test.js'],
+      {useJestCircus: true},
```
you can do `skipOnJasmine` instead, then it will run just with circus

SimenB replied:
https://github.com/facebook/jest/blob/d1ce3cd0135b5a8b64f84998fd5f36be4afb2bef/scripts/ConditionalTest.js#L17

palmerj3 replied:
Ah perfect thanks!

SimenB commented on docs/JestObjectAPI.md:
```diff
@@ -312,6 +313,21 @@ test('works too', () => {
 
 Returns the `jest` object for chaining.
 
+### `jest.retryTimes()`
```
should mention it just works with circus

palmerj3 replied:
I mention in that section that it works with jest-circus. Should I be more explicit?

SimenB replied:
Where is it mentioned?

palmerj3 replied:
Ah my mistake sorry. I mentioned it in the changelog not the docs. Will update.

SimenB commented on e2e/runJest.js:
```diff
@@ -21,6 +21,7 @@ const JEST_PATH = path.resolve(__dirname, '../packages/jest-cli/bin/jest.js');
 type RunJestOptions = {
   nodePath?: string,
   skipPkgJsonCheck?: boolean, // don't complain if can't find package.json
+  useJestCircus?: boolean,
```
we don't need the changes in this file now, right?

palmerj3 commented:
@aaronabramov @SimenB I think the PR is ready. I've addressed all the feedback so far.

But I'm getting a weird (unrelated?) error on some CI checks that I can't make sense of. All things pass locally on node 9 and node 10.

`TypeError: Cannot read property 'prototype' of undefined
`

Something related to graceful-fs

SimenB commented on e2e/__tests__/test_retries.test.js:
```diff
@@ -0,0 +1,96 @@
+      ],
+    };
+
+    // Test retries only available via JEST_CIRCUS
```
outdated comment, right?

palmerj3 replied:
Yep. Pushing a commit now to fix.

SimenB replied:
hmmm :P

palmerj3 replied:
All set - still getting CI errors due to the graceful-fs dependency bump yesterday (I believe)

SimenB replied:
The comment is still there :D

thymikee replied:
CI failures are not because of `graceful-fs` bump (there wasn't any afaik), but started to happen since Node 10.5

SimenB commented on e2e/runJest.js:
```diff
@@ -49,12 +49,15 @@ function runJest(
 
-  const env = options.nodePath
-    ? Object.assign({}, process.env, {
+  const nodePathOverrides = options.nodePath
```
needed change?

palmerj3 replied:
I can revert that back to the original

aaronabramov commented on e2e/snapshot-escape/__tests__/snapshot.test.js:
```diff
@@ -8,3 +8,4 @@
 
 // prettier-ignore
 test('escape strings', () => expect('one: \\\'').toMatchSnapshot());
+test('escape strings two', () => expect('two: \'"').toMatchSnapshot());
```
i think this line is not supposed to be committed.
i believe it's coming from our old janky snapshot tests that modifies our fixtures (that are under git and not in /tmp) and if you interrupt the test before it's done this thing will stay there :( (hence all other modifications to snapshots you had to make)

check `e2e/__tests__/snapshot.test.js`

palmerj3 replied:
Ah ok - should I remove this file or that line?

palmerj3 replied:
I saw other PRs failing with the same error

palmerj3 commented:
I really need to learn how to software :)  Outdated comments removed.

palmerj3 commented:
@aaronabramov so I removed that one test case but still getting the same error. Happy to revert that commit if you like. But if you have other ideas for things I can check I can do that too.

palmerj3 commented:
Think this is good to go now. Any more feedback?

palmerj3 commented:
@SimenB all set

palmerj3 commented:
Merge? :)

SimenB commented:
FB employees typically merge the larger changes :)

palmerj3 commented:
@aaronabramov any more thoughts on this? No rush on my end - would just rather not have to rebase again :)

aaronabramov commented:
it looks good to me!
